lib/vdsm/virt/vmdevices/storage.py

```diff
@@ -923,22 +925,54 @@ def get_snapshot_xml(self, snap_info):
 
     def on_block_threshold(self, reported_path):
         """
-        Callback to be executed when we receive a BLOCK_THRESHOLD
-        event from libvirt. We mark this drive accordingly,
-        so the periodic check will pick up this drive for
-        extension.
+        Called when disk allocation excceeds configured threshold.
+
+        Mark this drive for extension, to be picked by the periodic monitor
+        later.
+
+        Return True if threshold state was modified.
         """
         with self._lock:
             if reported_path != self._path:
                 self.log.debug(
                     "block threshold event mismatch drive %r path=%r "
                     "reported path=%r - ignored",
                     self.name, self._path, reported_path)
-                return
+                return False
+
+            if self._threshold_state == BLOCK_THRESHOLD.EXCEEDED:
+                self.log.debug(
+                    "drive %r block threshold already exceeded, ignored",
+                    self.name)
+                return False
+
+            self.log.info("drive %r block threshold exceeded", self.name)
+            self._threshold_state = BLOCK_THRESHOLD.EXCEEDED
+            self.exceeded_time = time.monotonic_time()
+            return True
+
+    def on_enospc(self):
+        """
+        Called when a VM pause because of ENOSPC writing to this drive.
+
+        We mark the drive for extension so it will be picked up by the priodic
+        monitor later.
+
+        Return True if threshold state was modified.
+        """
+        with self._lock:
+            if self._threshold_state == BLOCK_THRESHOLD.EXCEEDED:
+                self.log.debug(
+                    "drive %r block threshold already exceeded, ignored",
+                    self.name)
+                return False
 
+            self.log.info(
+                "drive %r force block threshold to exceeded",
+                self.name)
```

Comment:
Ha, moving it here may save some milliseconds from exceeded_time and make looking it faster! :-)

Reply:
Moving what here?

Reply:
self.exceeded_time is now assigned after logging rather than before logging (as it was before), so the tracked time will be slightly shorter.  No real issue, I suppose.

Reply:
OK, logging is fast and should never block, so it does not really matter if we log before or after.